libyul/backends/wasm/BinaryTransform.cpp

```diff
@@ -0,0 +1,549 @@
+	{"i64.load32_s", 0x34},
+	{"i64.load32_u", 0x35},
+	{"i32.store", 0x36},
+	{"i64.store", 0x37},
+	{"i32.store8", 0x3a},
+	{"i32.store16", 0x3b},
+	{"i64.store8", 0x3c},
+	{"i64.store16", 0x3d},
+	{"i64.store32", 0x3e},
+	{"memory.size", 0x3f},
+	{"memory.grow", 0x40},
+	{"i32.eqz", 0x45},
+	{"i32.eq", 0x46},
+	{"i32.ne", 0x47},
+	{"i32.lt_s", 0x48},
+	{"i32.lt_u", 0x49},
+	{"i32.gt_s", 0x4a},
+	{"i32.gt_u", 0x4b},
+	{"i32.le_s", 0x4c},
+	{"i32.le_u", 0x4d},
+	{"i32.ge_s", 0x4e},
+	{"i32.ge_u", 0x4f},
+	{"i64.eqz", 0x50},
+	{"i64.eq", 0x51},
+	{"i64.ne", 0x52},
+	{"i64.lt_s", 0x53},
+	{"i64.lt_u", 0x54},
+	{"i64.gt_s", 0x55},
+	{"i64.gt_u", 0x56},
+	{"i64.le_s", 0x57},
+	{"i64.le_u", 0x58},
+	{"i64.ge_s", 0x59},
+	{"i64.ge_u", 0x5a},
+	{"i32.clz", 0x67},
+	{"i32.ctz", 0x68},
+	{"i32.popcnt", 0x69},
+	{"i32.add", 0x6a},
+	{"i32.sub", 0x6b},
+	{"i32.mul", 0x6c},
+	{"i32.div_s", 0x6d},
+	{"i32.div_u", 0x6e},
+	{"i32.rem_s", 0x6f},
+	{"i32.rem_u", 0x70},
+	{"i32.and", 0x71},
+	{"i32.or", 0x72},
+	{"i32.xor", 0x73},
+	{"i32.shl", 0x74},
+	{"i32.shr_s", 0x75},
+	{"i32.shr_u", 0x76},
+	{"i32.rotl", 0x77},
+	{"i32.rotr", 0x78},
+	{"i64.clz", 0x79},
+	{"i64.ctz", 0x7a},
+	{"i64.popcnt", 0x7b},
+	{"i64.add", 0x7c},
+	{"i64.sub", 0x7d},
+	{"i64.mul", 0x7e},
+	{"i64.div_s", 0x7f},
+	{"i64.div_u", 0x80},
+	{"i64.rem_s", 0x81},
+	{"i64.rem_u", 0x82},
+	{"i64.and", 0x83},
+	{"i64.or", 0x84},
+	{"i64.xor", 0x85},
+	{"i64.shl", 0x86},
+	{"i64.shr_s", 0x87},
+	{"i64.shr_u", 0x88},
+	{"i64.rotl", 0x89},
+	{"i64.rotr", 0x8a},
+	{"i32.wrap_i64", 0xa7},
+	{"i64.extend_i32_s", 0xac},
+	{"i64.extend_i32_u", 0xad},
+};
+
+bytes opcode(Opcode _o)
+{
+	return bytes(1, uint8_t(_o));
+}
+
+bytes lebEncode(uint64_t _n)
+{
+	bytes encoded;
+	while (_n > 0x7f)
+	{
+		encoded.emplace_back(uint8_t(0x80 | (_n & 0x7f)));
+		_n >>= 7;
+	}
+	encoded.emplace_back(_n);
+	return encoded;
+}
+
+bytes lebEncodeSigned(int64_t _n)
+{
+	if (_n >= 0 && _n < 0x40)
+		return bytes(1, uint8_t(uint64_t(_n) & 0xff));
+	else if (-_n > 0 && -_n < 0x40)
+		return bytes(1, uint8_t(uint64_t(_n + 0x80) & 0xff));
+	else
+		return bytes(1, uint8_t(0x80 | uint8_t(_n & 0x7f))) + lebEncodeSigned(_n / 0x80);
+}
+
+}
+
+bytes BinaryTransform::run(Module const& _module)
+{
+	BinaryTransform bt;
+
+	for (size_t i = 0; i < _module.globals.size(); ++i)
+		bt.m_globals[_module.globals[i].variableName] = i;
+
+	size_t funID = 0;
+	for (FunctionImport const& fun: _module.imports)
+		bt.m_functions[fun.internalName] = funID++;
+	for (FunctionDefinition const& fun: _module.functions)
+		bt.m_functions[fun.name] = funID++;
+
+	bytes ret{0, 'a', 's', 'm'};
+	ret += bytes{1, 0, 0, 0};
+	ret += bt.typeSection(_module.imports, _module.functions);
+	ret += bt.importSection(_module.imports);
+	ret += bt.functionSection(_module.functions);
+	ret += bt.memorySection();
+	ret += bt.globalSection();
+	ret += bt.exportSection();
+	for (auto const& sub: _module.subModules)
+	{
+		// TODO should we prefix and / or shorten the name?
```

Comment:
Since the name is used as part of the custom section I'd opt for not making this unnecessarily long. Perhaps just consider replacing the name with a simple counter at some point?

Hm, actually, why not just merge all data into into a single custom section given all the access is via codecopy?

Reply:
Yeah, I would keep it like that for now, but good to keep it in mind for the future.